adb_graphics/datahandler/gribdata.py

```diff
@@ -481,128 +285,185 @@ class fieldData(UPPData):
+        flru = where((ceil > 0.0) & (ceil < 0.5), 3.01, flru)
+        flru = where((vis < 1.0), 3.01, flru)
 
         vis.close()
 
-        return xr.DataArray(flru)
+        return DataArray(flru)
 
     @property
     def cmap(self):
+        """
+        The LinearSegmentedColormap specified by the config key 'cmap'.
+        """
 
-        ''' Returns the LinearSegmentedColormap specified by the config key
-        "cmap" '''
-
-        return cm.get_cmap(self.vspec['cmap'])
+        return get_cmap(self.vspec["cmap"])
 
     @property
     def colors(self) -> np.ndarray:
-
-        '''
-        Returns a list of colors, specified by the config key "colors".
-
-        The yaml file "colors" key may contain a list or a function to be
-        called.
-        '''
-
-        color_spec = self.vspec.get('colors')
-
-        if isinstance(color_spec, (list, np.ndarray)):
-            return np.asarray(color_spec)
+        """
+        Returns an array of colors, specified by the config key "colors".
+        """
+
+        color_spec = self.vspec.get("colors", "")
+        if not color_spec:
+            msg = f"No colors definition found for {self.short_name} at {self.level}"
+            raise errors.NoGraphicsDefinitionForVariableError(msg)
         try:
             ret = self.__getattribute__(color_spec)
-            if callable(ret):
-                return ret()
-            return ret
-        except AttributeError:
-            return color_spec
+        except AttributeError as e:
+            msg = f"There is no color definition named {color_spec}"
+            raise AttributeError(msg) from e
+        if callable(ret):
+            return np.asarray(ret())
+        return np.asarray(ret)
 
     @property
     def corners(self) -> list:
+        """
 
-        '''
-        Returns lat and lon of lower left (ll) and upper right(ur) corners:
+        Returns lat and lon of lower left (ll) and upper right(ur) corners.
+
+        Order:
                ll_lat, ur_lat, ll_lon, ur_lon
-        '''
+        """
 
         lat, lon = self.latlons()
-        if self.model in ['global', 'hfip', 'obs']:
-            ret = [lat[-1], lat[0], lon[0], lon[-1]]
-        elif self.model == 'global_mpas':
-            ret = [lat[0], lat[-1], lon[0], lon[-1]]
-        else:
-            ret = [lat[0, 0], lat[-1, -1], lon[0, 0], lon[-1, -1]]
+        if len(lat.shape) == 2:
+            return [
+                np.round(x, decimals=6) for x in [lat[0, 0], lat[-1, -1], lon[0, 0], lon[-1, -1]]
+            ]
+        return [np.round(x, decimals=6) for x in [lat[0], lat[-1], lon[0], lon[-1]]]
 
-        return ret
+    @property
+    def data(self) -> DataArray:
+        """
+        Sets the data property on the object for use when we need to update
+        the values associated with a given object -- helpful for differences.
+        """
+        if not hasattr(self, "_data"):
+            self._data = self.values()
+        return self._data
```

Comment:
I'm wondering why not something like
```suggestion
    @cached_property
    def data(self) -> DataArray:
        """
        ???
        """
        return self.values()
```
If I'm reading the original right, this is a one-time thing: `_data` will not be updated even if whatever is returned by `.values()` changes? If so, wouldn't a cached property provide the same, unless you specifically need to access the `._data` property elsewhere?

But I also don't understand the docstring: Is this mean to support `dict`-like calling of `.update()` on instances of this class? I don't immediately see how this would let one "update the values", at least on _this_ object.

Reply:
I probably don't have the full getter/setter pattern down pat here, but I found that a `cached_property` would work for that. I'm using the word `update` in the original comment interchangeably with `set`. Elsewhere in the code, this value may be set to the difference between the values of the original object and the values of a second object when we do difference maps. If I use a `cached_property` here, the diff doesn't actually propagate after setting the value externally.

Reply:
I'm probably not going to understand the full range of functionality at the moment, but I found this confusing. If I'd seen two methods, decorated with `@property` and `@data.setter`, I'd have recognized this as a getter/setter patern and then I don't think any docstring would be necessary, as it'd be obvious.

Given that this isn't a library exposed to the outside world, could you just have
```
self.data = self.values()
```
in `__init__()`? The getter/setter pattern is useful when you need to constrain the behavior or users, but you're your own user here.

Reply:
I feel like perhaps pylint sent me down this rabbit hole in the past. Let me see if ruff is more forgiving.

Reply:
Oooh! ruff does not care!

Reply:
I think I'm going to need to leave this for now. Without the setter/getter pattern, `mypy` complains that I'm trying to set a read-only value when I do that from other places in the code. I also run into issues being able to find some of the transform methods associated with this class when I call `self.get_values()` (used to be `self.values()`) from the `__init__` method. I will table this for future work.

Reply:
Sure, but just in case I wasn't clear, in the end I was wondering why any kind of method wrapper around `data` is necessary, and why `data` can't just be an instance variable. Maybe that _was_ clear, just making sure.